staging/src/slime.io/slime/modules/meshregistry/pkg/source/zookeeper/watching.go

```diff
@@ -32,228 +35,498 @@ func (s *Source) ServiceNodeDelete(path string) {
+}
+
+func (w *worker) HandleEvent(e ServiceEvent) {
+	w.q.Push(e)
+}
+
+func (w *worker) process(ctx context.Context) {
+	for w.processItem(ctx) {
 	}
+}
 
-	sort.Slice(keys, func(i, j int) bool {
-		return keys[i] < keys[j]
-	})
-	var buff bytes.Buffer
-	for _, k := range keys {
-		buff.WriteString(k)
-		buff.WriteString(labels[k])
+func (w *worker) processItem(ctx context.Context) bool {
+	select {
+	case <-ctx.Done():
+		return false
+	default:
 	}
-	return buff.String()
+	e := w.q.Pop()
+	watcher, ok := w.cache.Get(e.path)
+	if e.etype == EventTypeDelete && ok {
+		ew := watcher.(*EndpointWatcher)
+		ew.Exit() // may block other service events from being processed？
+		scope.Infof("endpoint watcher %q exited", ew.servicePath)
+		w.cache.Remove(e.path)
+		return true
+	}
+	if e.etype == EventTypeCreate && !ok {
+		ew := NewEndpointWatcher(e.path, w.epWatcherOpts)
+		w.cache.Set(e.path, ew)
+		scope.Infof("service worker start a new endpoint watcher for %s", e.path)
+		ew.Start(ctx)
+		return true
+	}
+	return true
+}
+
+var dubboExcludeServicePath = []string{
+	"metadata", // The application level metadata of zookeeper is located in /dubbo/metadata
+	"mapping",  // The interface-application_name mapping information of zookeeper is located in /dubbo/mapping
+}
+
+type ServiceWatcher struct {
+	ctx context.Context
+
+	conn               *atomic.Value
+	rootPath           string
+	endpointUpdateFunc func([]string, []string, string)
+	serviceDeleteFunc  func(string)
+	gatewatModel       bool
+
+	svcs []string
+
+	workers []*worker
+
+	initLock               sync.Mutex
+	initWait               sync.WaitGroup
+	initCnt, initThreshold int
 }
 
-func sortEndpoint(endpoints []*v1alpha3.WorkloadEntry) {
-	sort.Slice(endpoints, func(i, j int) bool {
-		if endpoints[i].Address == endpoints[j].Address {
-			return labelToSstring(endpoints[i].Labels) < labelToSstring(endpoints[j].Labels)
+// block until initialized
+func (sw *ServiceWatcher) Start(ctx context.Context) {
+	sw.ctx = ctx
+	initCh := make(chan struct{})
+	go func() {
+		firstLoop := true
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+			}
+
+			paths, _, e, err := sw.conn.Load().(*zk.Conn).ChildrenW(sw.rootPath)
+			if err != nil {
+				if errors.Is(err, zk.ErrNoNode) {
+					// mark ready
+					firstLoop = false
+					close(initCh)
+				}
+				continue
+			}
+			svcs := make([]string, 0, len(paths))
+			for _, s := range paths {
+				for _, exclude := range dubboExcludeServicePath {
+					if s != exclude {
+						svcs = append(svcs, s)
+					}
+				}
+			}
+			if firstLoop {
+				firstLoop = false
+				// For the zk watching mode, we initialize according to the number of services
+				// obtained after registering the watch root node for the first time.
+				// When the ep watchers for these services start up and perform their first processing,
+				// the zk source is marked as ready.
+				initThreshold := len(svcs)
+				sw.initThreshold = initThreshold
```

Comment:
If the number of services is reduced from 1000 to 999, and the service event corresponding to the deleted service has not been processed, it may be merged away without the corresponding ep watcher done operation, then it will wait forever.